### PR TITLE
Callbacks perf

### PR DESCRIPTION
**Pull request by ggazzo**
I used:
```javascript
const combine = (f, x) => (e, ...constants) => x(f(e, ...constants), ...constants);
```

to create a single function to avoid `.reduce` each time we invoke a callback.

I set a name for each callback to help Prometheus to show the callbacks name properly

I removed the timers, and created a wrappers IF we want to use, if not we dont need track this time anymore.

### CODE REVIEW

File: app/callbacks/lib/callbacks.js [hunk 1]
```diff
@@ -2,22 +2,64 @@ import { Meteor } from 'meteor/meteor';
 import { Random } from 'meteor/random';
 import _ from 'underscore';
 
+let timed = false;
+
+if (Meteor.isClient) {
+	const { getConfig } = require('../../ui-utils/client/config');
+	timed = [getConfig('debug'), getConfig('timed-callbacks')].includes('true');
```
Comment (sampaiodiego) [inline app/callbacks/lib/callbacks.js]:
```suggestion
	timed = [getConfig('debug'), getConfig('timed-callbacks')].includes('true');
```

File: app/callbacks/lib/callbacks.js [hunk 2]
```diff
@@ -2,22 +2,64 @@ import { Meteor } from 'meteor/meteor';
...
+const wrapRun = (hook, fn) => (...args) => {
+	const time = Date.now();
+	const ret = fn(...args);
+	const totalTime = Date.now() - time;
+	console.log(`${ hook }:`, totalTime);
+	return ret;
+};
+
+const handleResult = (fn) => {
+	const ret = (result, constant) => {
```
Comment (sampaiodiego) [inline app/callbacks/lib/callbacks.js]:
```suggestion
	return (result, constant) => {
```

File: app/callbacks/lib/callbacks.js [hunk 3]
```diff
@@ -2,22 +2,64 @@ import { Meteor } from 'meteor/meteor';
...
+		const callbackResult = callbacks.runItem({ callback: fn, result, constant });
+		return typeof callbackResult === 'undefined' ? result : callbackResult;
+	};
+	return ret;
+};
+
+
+const empty = (e) => e;
```
Comment (sampaiodiego) [inline app/callbacks/lib/callbacks.js]: :-1:

File: app/callbacks/lib/callbacks.js [hunk 4]
```diff
@@ -124,12 +143,10 @@ callbacks.run = function(hook, item, constant) {
...
 */
 
-callbacks.runAsync = function(hook, item, constant) {
-	const callbackItems = callbacks[hook];
-	if (Meteor.isServer && callbackItems && callbackItems.length) {
-		Meteor.defer(function() {
-			callbackItems.forEach((callback) => callback(item, constant));
-		});
+callbacks.runAsync = Meteor.isServer ? function(hook, item, constant) {
+	const callbackItems = combinedCallbacks[hook];
```
Comment (sampaiodiego) [inline app/callbacks/lib/callbacks.js]:
```suggestion
	const callbackItems = callbacks[hook];
```

File: app/graphql/server/resolvers/messages/chatMessageAdded.js [hunk 5]
```diff
@@ -44,7 +44,7 @@ const resolver = {
 
 callbacks.add('afterSaveMessage', (message) => {
 	publishMessage(message);
-}, null, 'chatMessageAddedSubscription');
+}, callbacks.priority.MEDIUM, 'joinDiscussionOnMessage', 'chatMessageAddedSubscription');
```
Comment (sampaiodiego) [inline app/graphql/server/resolvers/messages/chatMessageAdded.js]:
```suggestion
}, callbacks.priority.MEDIUM, 'chatMessageAddedSubscription');
```

File: app/search/server/events/events.js [hunk 6]
```diff
@@ -24,11 +24,11 @@ const eventService = new EventService();
  */
 callbacks.add('afterSaveMessage', function(m) {
 	eventService.promoteEvent('message.save', m._id, m);
-});
+}, callbacks.priority.MEDIUM, 'search-events');
 
 callbacks.add('afterDeleteMessage', function(m) {
 	eventService.promoteEvent('message.delete', m._id);
-});
+}, callbacks.priority.MEDIUM, 'search-events');
```
Comment (sampaiodiego) [inline app/search/server/events/events.js]:
```suggestion
}, callbacks.priority.MEDIUM, 'search-events-delete');
```

File: app/lib/server/functions/sendMessage.js [hunk 7]
```diff
@@ -211,7 +211,7 @@ export const sendMessage = function(user, message, room, upsert = false) {
 		Defer other updates as their return is not interesting to the user
 		*/
 		// Execute all callbacks
-		Meteor.defer(() => callbacks.run('afterSaveMessage', message, room, user._id));
+		callbacks.run('afterSaveMessage', message, room, user._id);
```
Comment (sampaiodiego) [inline app/lib/server/functions/sendMessage.js]:
```suggestion
		Meteor.defer(() => callbacks.run('afterSaveMessage', message, room, user._id));
```

please revert

File: app/callbacks/lib/callbacks.js [hunk 8]
```diff
@@ -2,22 +2,61 @@ import { Meteor } from 'meteor/meteor';
...
+};
+
+const handleResult = (fn) => (result, constant) => {
+	const callbackResult = callbacks.runItem({ hook: fn.hook, callback: fn, result, constant });
+	return typeof callbackResult === 'undefined' ? result : callbackResult;
+};
 
 
+const empty = (e) => e;
+const combine = (f, x) => (e, ...constants) => x(f(e, ...constants), ...constants);
```
Comment (tassoevan) [inline app/callbacks/lib/callbacks.js]:
```suggestion
const pipe = (f, g) => (e, ...constants) => g(f(e, ...constants), ...constants);
```

File: app/callbacks/lib/callbacks.js [hunk 9]
```diff
@@ -2,22 +2,61 @@ import { Meteor } from 'meteor/meteor';
...
+const createCallback = (hook, callbacks) => callbacks.map(handleResult).reduce(combine, empty);
```
Comment (tassoevan) [inline app/callbacks/lib/callbacks.js]:
```suggestion
const createCallback = (hook, callbacks) => callbacks.map(handleResult).reduce(pipe, identity);
```

File: app/callbacks/lib/callbacks.js [hunk 10]
```diff
@@ -2,22 +2,61 @@ import { Meteor } from 'meteor/meteor';
...
+
+const createCallbackTimed = (hook, callbacks) =>
+	wrapRun(hook,
+		callbacks
+			.map(wrapCallback)
+			.map(handleResult)
+			.reduce(combine, empty)
```
Comment (tassoevan) [inline app/callbacks/lib/callbacks.js]:
```suggestion
			.reduce(pipe, identity)
```

**Pull request merged**